### PR TITLE
new balancing algorithm + misc fixes

### PR DESCRIPTION
**Pull request by lexikiq**
- balancing now automatically determines balance percentages if not set (by spreading evenly among teams (without percentages)), ensures teams meet minimum player requirements, and appropriately balances players who join late

  - (Un)balancedTeamMechanics have been removed, as all they did was implement different balancing algorithms which have been obsoleted

- late join now works on unbalanced team games (oops)

- murder now ends properly when players are spectating

- deathmatch code was repetitive so i made a common mechanic for it

- players are now automatically balanced to another team on death if teams are unbalanced

closes Pugabyte/BearNation#549

### CODE REVIEW

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java [hunk 1]
```diff
@@ -201,4 +207,214 @@ public void onQuit(MatchQuitEvent event) {
...
+	@Override
+	public void onDeath(MinigamerDeathEvent event) {
+		// auto-balancing
+		Match match = event.getMatch();
+		Minigamer minigamer = event.getMinigamer();
+		super.onDeath(event);
+		if (!minigamer.isAlive() || match.isEnded())
+			return;
+
+		List<BalanceWrapper> wrappers = getBalanceWrappers(match).stream().filter(w -> !w.getTeam().equals(minigamer.getTeam()) && w.percentageDiscrepancy() > 0 && w.extraPlayerPercentDiscrepancy() >= 0).collect(Collectors.toList());
```
Comment (Pugabyte) [inline src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:270]: Please try to never use single character variable names

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java [hunk 2]
```diff
@@ -201,4 +207,214 @@ public void onQuit(MatchQuitEvent event) {
...
+		if (wrappers.isEmpty())
+			return;
+		// sort teams by closest to being equal
+		wrappers.sort((w1, t1) -> (int) ((w1.extraPlayerPercentDiscrepancy() - t1.extraPlayerPercentDiscrepancy())*100));
+		// select randomly if multiple teams are equal
+		List<BalanceWrapper> randomWrappers = new ArrayList<>();
+		randomWrappers.add(wrappers.get(0));
+		double val = wrappers.get(0).extraPlayerPercentDiscrepancy();
+		int c = 1;
```
Comment (Pugabyte) [inline src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:279]: Same as above

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java [hunk 3]
```diff
@@ -201,4 +207,214 @@ public void onQuit(MatchQuitEvent event) {
...
+		while (!minigamers.isEmpty()) {
+			// this basically finds the team with the largest percent
+			wrappers = wrappers.stream().filter(wrapper -> wrapper.getNeededPlayers() != -1).sorted().collect(Collectors.toList());
+			if (wrappers.isEmpty())
+				break;
+			// get teams with matching percentage discrepancies (ie the teams are perfectly balanced) and randomly
+			//  select one of them
+			List<BalanceWrapper> equalWrappers = new ArrayList<>();
+			equalWrappers.add(wrappers.get(0));
+			Nexus.severe(String.valueOf(wrappers.get(0).extraPlayerPercentDiscrepancy()));
```
Comment (Pugabyte) [inline src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:330]: Leftover debug?

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java [hunk 4]
```diff
@@ -201,4 +207,214 @@ public void onQuit(MatchQuitEvent event) {
...
+		Minigamer minigamer = event.getMinigamer();
+		super.onDeath(event);
+		if (!minigamer.isAlive() || match.isEnded())
+			return;
+
+		List<BalanceWrapper> wrappers = getBalanceWrappers(match).stream().filter(w -> !w.getTeam().equals(minigamer.getTeam()) && w.percentageDiscrepancy() > 0 && w.extraPlayerPercentDiscrepancy() >= 0).collect(Collectors.toList());
+		if (wrappers.isEmpty())
+			return;
+		// sort teams by closest to being equal
+		wrappers.sort((w1, t1) -> (int) ((w1.extraPlayerPercentDiscrepancy() - t1.extraPlayerPercentDiscrepancy())*100));
```
Comment (Pugabyte) [inline src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:274]: Is this sorting not already implemented in the comparable override?

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java [hunk 5]
```diff
@@ -201,4 +207,214 @@ public void onQuit(MatchQuitEvent event) {
...
+
+		Match match = minigamers.get(0).getMatch();
+		Arena arena = match.getArena();
+		List<Team> teams = new ArrayList<>(arena.getTeams());
+
+		int required = 0;
+		for (Team team : teams) required += team.getMinPlayers();
+
+		if (match.getMinigamers().size() < required) {
+			criticalErrorAbort("Not enough players to meet team requirements!", match);
```
Comment (Pugabyte) [inline src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:222]: Can this be renamed to just `error` to match the commands framework

**Review by Pugabyte — CHANGES_REQUESTED** [review]
Some minor requests, overall very good PR, I'm impressed with your java/coding knowledge. As long as its been tested thoroughly, I'll approve after the below comments are addressed :)

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java (hunk 4 above)
Comment (lexikiq) [reply src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:274]: the comparable sorts by `.percentageDiscrepancy()` and in the other direction, but you're right, you'd get the same results by using the comparable and reversing the list

File: src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java (hunk 3 above)
Comment (lexikiq) [reply src/main/java/me/pugabyte/nexus/features/minigames/models/mechanics/multiplayer/teams/TeamMechanic.java:330]: ah yeah whoops

**Comment by lexikiq** [comment]
> overall very good PR, I'm impressed with your java/coding knowledge
aww thank you :) code has been tested as hard as i can with me and my 4 alts, should all be good!

**Pull request merged**